File: pb/pb_hooks/exif.js
// Strip GPS / EXIF so a public field photo cannot publish a child's weekend location.
// Keep in lockstep with lib/exif.py.

function toBytes(raw) {
  if (!raw) return [];
  if (typeof raw === "string") {
    const out = [];
    for (let i = 0; i < raw.length; i++) out.push(raw.charCodeAt(i) & 255);
    return out;
  }
  if (raw.length !== undefined) {
    const out = [];
    for (let i = 0; i < raw.length; i++) out.push(Number(raw[i]) & 255);
    return out;
  }
  return [];
}

function fromBytes(bytes) {
  let s = "";
  for (let i = 0; i < bytes.length; i++) s += String.fromCharCode(bytes[i]);
  return s;
}

function stripJpeg(bytes) {
  if (bytes.length < 4 || bytes[0] !== 0xff || bytes[1] !== 0xd8) return bytes;
  const out = [0xff, 0xd8];
  let i = 2;
  while (i < bytes.length - 1) {
    if (bytes[i] !== 0xff) {
      for (; i < bytes.length; i++) out.push(bytes[i]);
      break;
    }
    const marker = bytes[i + 1];
    if (marker === 0xda) {
      for (; i < bytes.length; i++) out.push(bytes[i]);
      break;
    }
    if (marker === 0xd9) {
      out.push(0xff, 0xd9);
      break;
    }
    if (i + 3 >= bytes.length) break;
    const len = (bytes[i + 2] << 8) | bytes[i + 3];
    const skip = marker === 0xe1 || marker === 0xe2;
    if (!skip) {
      for (let j = 0; j < len + 2 && i + j < bytes.length; j++) out.push(bytes[i + j]);
    }
    i += len + 2;
  }
  return out;
}

function pngChunkType(bytes, i) {
  return String.fromCharCode(bytes[i], bytes[i + 1], bytes[i + 2], bytes[i + 3]);
}

function stripPng(bytes) {
  const sig = [137, 80, 78, 71, 13, 10, 26, 10];
  if (bytes.length < 16) return bytes;
  for (let i = 0; i < 8; i++) if (bytes[i] !== sig[i]) return bytes;
  const out = sig.slice();
  let i = 8;
  while (i + 12 <= bytes.length) {
    const len = (bytes[i] << 24) | (bytes[i + 1] << 16) | (bytes[i + 2] << 8) | bytes[i + 3];
    const type = pngChunkType(bytes, i + 4);
    const total = 12 + len;
    if (i + total > bytes.length) break;
    const drop = type === "eXIf" || type === "iTXt" || type === "tEXt" || type === "zTXt";
    if (!drop) {
      for (let j = 0; j < total; j++) out.push(bytes[i + j]);
    }
    i += total;
    if (type === "IEND") break;
  }
  return out;
}

function stripExif(raw) {
  const bytes = toBytes(raw);
  if (bytes.length >= 3 && bytes[0] === 0xff && bytes[1] === 0xd8) return stripJpeg(bytes);
  if (bytes.length >= 8 && bytes[0] === 137 && bytes[1] === 80) return stripPng(bytes);
  return bytes;
}

function hasGpsExif(raw) {
  const bytes = toBytes(raw);
  if (bytes.length >= 3 && bytes[0] === 0xff && bytes[1] === 0xd8) {
    const text = fromBytes(bytes);
    return text.indexOf("GPS") !== -1 || text.indexOf("Exif") !== -1;
  }
  if (bytes.length >= 8 && bytes[0] === 137) {
    const text = fromBytes(bytes);
    return text.indexOf("eXIf") !== -1;
  }
  return false;
}

module.exports = {
  stripExif: stripExif,
  hasGpsExif: hasGpsExif,
  toBytes: toBytes,
};

File: lib/exif.py
"""Strip JPEG APP1/APP2 and PNG text/EXIF chunks. Lockstep with pb/pb_hooks/exif.js."""

from __future__ import annotations


def strip_jpeg(data: bytes) -> bytes:
    if len(data) < 4 or data[:2] != b"\xff\xd8":
        return data
    out = bytearray(b"\xff\xd8")
    i = 2
    while i < len(data) - 1:
        if data[i] != 0xFF:
            out.extend(data[i:])
            break
        marker = data[i + 1]
        if marker == 0xDA:
            out.extend(data[i:])
            break
        if marker == 0xD9:
            out.extend(b"\xff\xd9")
            break
        if i + 3 >= len(data):
            break
        length = (data[i + 2] << 8) | data[i + 3]
        skip = marker in (0xE1, 0xE2)
        if not skip:
            out.extend(data[i : i + length + 2])
        i += length + 2
    return bytes(out)


def strip_png(data: bytes) -> bytes:
    sig = b"\x89PNG\r\n\x1a\n"
    if not data.startswith(sig):
        return data
    out = bytearray(sig)
    i = 8
    while i + 12 <= len(data):
        length = int.from_bytes(data[i : i + 4], "big")
        typ = data[i + 4 : i + 8]
        total = 12 + length
        if i + total > len(data):
            break
        if typ not in (b"eXIf", b"iTXt", b"tEXt", b"zTXt"):
            out.extend(data[i : i + total])
        i += total
        if typ == b"IEND":
            break
    return bytes(out)


def strip_exif(data: bytes) -> bytes:
    if data[:2] == b"\xff\xd8":
        return strip_jpeg(data)
    if data[:8] == b"\x89PNG\r\n\x1a\n":
        return strip_png(data)
    return data
